cpp/FileMetadataInitializer.cpp

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2018 Snowflake Computing, Inc. All rights reserved.
+ */
+
+#include "FileMetadataInitializer.hpp"
+#include "EncryptionProvider.hpp"
+#include <dirent.h>
+#include <fnmatch.h>
+
+// used to decide whether to upload in sequence or in parallel
+#define DATA_SIZE_THRESHOLD 5242880
+
+Snowflake::Client::FileMetadataInitializer::FileMetadataInitializer(
+  std::vector<FileMetadata> * smallFileMetadata,
+  std::vector<FileMetadata> * largeFileMetadata) :
+  m_smallFileMetadata(smallFileMetadata),
+  m_largeFileMetadata(largeFileMetadata),
+  m_autoCompress(true)
+{
+}
+
+void Snowflake::Client::FileMetadataInitializer::initFileMetadata(
+  std::string &fileDir, char *fileName)
+{
+  std::string srcFileName = fileDir + fileName;
+  struct stat fileStatus;
+  if (!stat(srcFileName.c_str(), &fileStatus))
+  {
+    if (S_ISREG(fileStatus.st_mode))
+    {
+      std::vector<FileMetadata> *metaListToPush =
+        fileStatus.st_size > DATA_SIZE_THRESHOLD ?
+        m_largeFileMetadata : m_smallFileMetadata;
+
+      metaListToPush->emplace_back();
+      metaListToPush->back().srcFileName = srcFileName;
+      metaListToPush->back().srcFileSize = (long)fileStatus.st_size;
+      metaListToPush->back().destFileName = std::string(fileName);
+
+      // process compression type
+      initCompressionMetadata(&metaListToPush->back());
+    }
+  }
+  else
+  {
+    throw;
+  }
+}
+
+void Snowflake::Client::FileMetadataInitializer::populateSrcLocMetadata(
+  std::string &sourceLocation)
+{
+  unsigned long dirSep = sourceLocation.find_last_of('/');
+  std::string dirPath = sourceLocation.substr(0, dirSep + 1);
+  std::string filePattern = sourceLocation.substr(dirSep + 1);
+
+  DIR * dir = nullptr;
+  struct dirent * dir_entry;
+  if ((dir = opendir(dirPath.c_str())) != NULL)
+  {
+    while((dir_entry = readdir(dir)) != NULL)
+    {
+      if (!fnmatch(filePattern.c_str(), dir_entry->d_name, 0))
+      {
+        initFileMetadata(dirPath, dir_entry->d_name);
+      }
+    }
+    closedir(dir);
+  }
+  else
+  {
+    // open dir failed
+    throw;
+  }
+}
+
+void Snowflake::Client::FileMetadataInitializer::initCompressionMetadata(
+  FileMetadata *fileMetadata)
+{
+  if(!strncasecmp(m_sourceCompression, "AUTO_DETECT", 11) ||
+     !strncasecmp(m_sourceCompression, "AUTO", 4))
+  {
+    // guess
+    fileMetadata->sourceCompression = FileCompressionType::guessCompressionType(
+      fileMetadata->srcFileName);
+  }
+  else if (!strncasecmp(m_sourceCompression, "NONE", 4))
+  {
+    fileMetadata->sourceCompression = &FileCompressionType::NONE;
+  }
+  else
+  {
+    // look up
+    fileMetadata->sourceCompression = FileCompressionType::lookUpByName(
+      m_sourceCompression);
+
+    if (!fileMetadata->sourceCompression)
+    {
+      // no compression found
+      throw;
+    }
+  }
+
+  if (fileMetadata->sourceCompression == &FileCompressionType::NONE)
+  {
+    fileMetadata->targetCompression = &FileCompressionType::GZIP;
```

Comment:
Is this the similar one I recently fixed in the Python Connector? If no auto compress, then the file compression type would not be Gzip? I'm not sure what this gzip mean.

Reply:
This GZIP is the data returned to the caller i.e. user will know that target compression is gzip. It will not influence the upload result since s3 connector did not need to know the compression type of file. The bug in Python Connector is because the upload metadata contains a content-encoding metadata indicating the source is gzip compression. Any way, will fix it.